File: cpp/tensorrt_llm/batch_manager/handleContextLogits.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "tensorrt_llm/batch_manager/handleContextLogits.h"

#include "tensorrt_llm/batch_manager/decoderBuffers.h"
#include "tensorrt_llm/batch_manager/llmRequest.h"
#include "tensorrt_llm/batch_manager/medusaBuffers.h"
#include "tensorrt_llm/batch_manager/runtimeBuffers.h"
#include "tensorrt_llm/common/nvtxUtils.h"
#include "tensorrt_llm/runtime/iTensor.h"
#include "tensorrt_llm/runtime/runtimeKernels.h"
#include "tensorrt_llm/runtime/utils/debugUtils.h"

namespace tr = tensorrt_llm::runtime;
namespace tru = tensorrt_llm::runtime::utils;

namespace tensorrt_llm::batch_manager
{

using BufferManager = tensorrt_llm::runtime::BufferManager;
using TensorPtr = runtime::ITensor::SharedPtr;
using ITensor = runtime::ITensor;
using SizeType32 = tensorrt_llm::runtime::SizeType32;

namespace
{

//! @brief Copy logits from context phase to beginning of generation logits.
//! @details Usually, this concerns logits of 1 token. In speculative decoding this concerns draftLen + 1 tokens.
void copyLastContextLogits(TensorPtr const& contextLogits, LlmRequest& llmReq, BufferManager const& bufferManager)
{
    TLLM_LOG_TRACE("%s start", __PRETTY_FUNCTION__);
    auto const numLogits = contextLogits->getShape().d[0];
    for (int beam = 0; beam < llmReq.getBeamWidthByIter(); beam++)
    {
        // [beamWidth, mMaxNewTokens, vocabSizePadded] -> [numLogits, vocabSizePadded]
        auto beamHostTensorPtr = ITensor::slice(llmReq.getGenerationLogitsHost(), {beam, 0}, numLogits);
        bufferManager.copy(*contextLogits, *beamHostTensorPtr);
    }
    TLLM_LOG_TRACE("%s stop", __PRETTY_FUNCTION__);
}

void setupMedusaLogits(std::vector<TensorPtr>& medusaLogitsHeads, TensorPtr const& medusaLogitsDevice,
    SizeType32 medusaHeads, SizeType32 logitsIndex, SizeType32 numLogits)
{
    for (SizeType32 hi = 0; hi < medusaHeads; ++hi)
    {
        TensorPtr logitsHead = ITensor::slice(medusaLogitsDevice, hi, 1);
        logitsHead->squeeze(0);
        medusaLogitsHeads[hi] = ITensor::slice(logitsHead, logitsIndex, numLogits);
    }
}

} // namespace

SizeType32 HandleContextLogits::operator()(DecoderInputBuffers& inputBuffers, RequestVector const& contextRequests,
    tr::ITensor::SharedPtr const& logits, std::vector<tr::SizeType32> const& numContextLogitsVec,
    tr::ModelConfig const& modelConfig, tr::BufferManager const& manager,
    OptionalRef<MedusaBuffers> medusaBuffers) const
{
    TLLM_LOG_TRACE("%s start", __PRETTY_FUNCTION__);
    NVTX3_SCOPED_RANGE(HandleContextLogits);

    auto& decoderRequests = inputBuffers.decoderRequests;
    decoderRequests.clear();
    decoderRequests.reserve(contextRequests.size());
    auto& allDecoderLogits = inputBuffers.logits;
    allDecoderLogits.clear();
    allDecoderLogits.reserve(contextRequests.size());

    SizeType32 batchIndex{0};
    SizeType32 logitsIndex{0};
    // Copy logits into decoderBuffers.logits
    for (auto const& llmReq : contextRequests)
    {
        auto const numContextLogits = numContextLogitsVec.at(batchIndex);
        auto const draftLength = llmReq->isLastContextChunk() ? llmReq->getNumDraftTokens() : 0;

        TLLM_LOG_DEBUG("logitsIndex: %d", logitsIndex);
        TLLM_LOG_DEBUG("numContextLogits %d", numContextLogits);
        TLLM_LOG_DEBUG("draftLength: %d", draftLength);

        if (modelConfig.computeContextLogits())
        {
            // Since the computational graph has been modified, only the last token is needed.
            TLLM_CHECK_WITH_INFO(!modelConfig.getSpeculativeDecodingMode().isMedusa()
                    && !modelConfig.getSpeculativeDecodingMode().isLookaheadDecoding(),
                "Return context logits is not supported with Medusa and Lookahead decoding");

            if (llmReq->getReturnContextLogits())
            {
                if (llmReq->getPrepopulatedPromptLen() > 0)
                {
                    TLLM_LOG_WARNING(
                        "Because of KV cache reuse, not all context logits could be produced for request %lu.",
                        llmReq->mRequestId);
                }
                TensorPtr contextLogitsDeviceView = ITensor::slice(logits, logitsIndex, numContextLogits);
                TensorPtr contextLogitsHostView = ITensor::slice(
                    llmReq->getContextLogitsHost(), llmReq->getContextCurrentPosition(), numContextLogits);
                // Copy to host directly
                manager.copy(*contextLogitsDeviceView, *contextLogitsHostView);
            }
        }
        logitsIndex += numContextLogits + draftLength;

        // Get the logits from the last context token and draft tokens
        auto const numDecoderLogits = 1 + draftLength;
        auto const seqSlot = llmReq->mSeqSlot.value();
        TensorPtr logitsView = ITensor::slice(logits, logitsIndex - numDecoderLogits, numDecoderLogits);

        if (modelConfig.getSpeculativeDecodingMode().hasDraftLogits())
        {
            auto& medusaLogitsHeads = inputBuffers.predictedDraftLogits.at(seqSlot);
            TLLM_CHECK(medusaBuffers);
            setupMedusaLogits(medusaLogitsHeads, medusaBuffers->medusaLogitsDevice,
                modelConfig.getSpeculativeDecodingModule().getMaxDraftPathLen(), logitsIndex - numDecoderLogits,
                numDecoderLogits);
        }

        // Save the last token logits of context into generation logits or
        // save the accepted token logits from target model
        if (llmReq->getReturnGenerationLogits())
        {
            copyLastContextLogits(logitsView, *llmReq, manager);
        }

        TLLM_CHECK_DEBUG_WITH_INFO(tru::tensorHasInvalid<float>(*logitsView, manager, "logits") == false,
            "Found invalid number (NaN or Inf) in logits");

        if (llmReq->isLastContextChunk())
        {
            TensorPtr decoderLogits;
            auto const reqBeamWidth = llmReq->getBeamWidthByIter();
            if (reqBeamWidth > 1)
            {
                // Tile logits of context requests
                auto const& logitsShape = logitsView->getShape();
                auto const logitsType = logitsView->getDataType();
                decoderLogits = manager.gpu(ITensor::makeShape({reqBeamWidth, logitsShape.d[1]}), logitsType);
                tensorrt_llm::runtime::kernels::tileTensor(
                    *decoderLogits, *logitsView, reqBeamWidth, manager.getStream());
                decoderLogits->unsqueeze(0);
            }
            else
            {
                decoderLogits = logitsView;
                decoderLogits->unsqueeze(1);
            }
            decoderRequests.push_back(llmReq);
            allDecoderLogits.emplace_back(std::move(decoderLogits));
        }

        ++batchIndex;
    }

    TLLM_LOG_TRACE("%s stop", __PRETTY_FUNCTION__);
    return logitsIndex;
}

} // namespace tensorrt_llm::batch_manager

File: tensorrt_llm/_torch/pyexecutor/handle_context_logits.py
from typing import List

import torch

from tensorrt_llm._torch.pyexecutor.llm_request import LlmRequest
from tensorrt_llm.bindings.internal.batch_manager import DecoderBuffers
from tensorrt_llm.logger import logger


class HandleContextLogits:

    def __call__(self, context_requests: List[LlmRequest],
                 num_context_logits_vec: List[int], logits: torch.Tensor,
                 decoder_buffers: DecoderBuffers) -> int:
        """Handle context logits for a batch of requests.

        Args:
            context_requests: List of context requests to process
            num_context_logits_vec: Number of context logits for each request
            logits: Input logits tensor
            decoder_buffers: Decoder buffers for storing intermediate results

        Returns:
            int: Index into logits tensor after processing all requests
        """
        logits_index = 0

        # Copy logits into decoderBuffers.logits
        decoder_buffer_logits = [torch.empty(0)] * len(decoder_buffers.logits)
        for batch_index, llm_req in enumerate(context_requests):
            num_context_logits = num_context_logits_vec[batch_index]
            draft_length = llm_req.num_draft_tokens if llm_req.is_last_context_chunk(
            ) else 0

            if llm_req.py_return_context_logits:
                if llm_req.prepopulated_prompt_len > 0:
                    logger.warning(
                        f"Because of KV cache reuse, not all context logits could be produced for request {llm_req.request_id}."
                    )

                context_logits_device_view = logits[logits_index:logits_index +
                                                    num_context_logits]
                llm_req.py_result.append_context_logits(
                    context_logits_device_view)

            logits_index += num_context_logits + draft_length

            # Get the logits from the last context token and draft tokens
            num_decoder_logits = 1 + draft_length
            seq_slot = llm_req.seq_slot
            logits_view = logits[logits_index - num_decoder_logits:logits_index]

            # Create a view of logits_view with shape (logits_view.shape[0], 1, logits_view.shape[1])
            # This creates a new tensor that shares the same underlying data
            decoder_buffer_logits[seq_slot] = logits_view.reshape(
                logits_view.shape[0], 1, logits_view.shape[1])

            # Save the last context token logits in generation logits storage
            if llm_req.py_return_generation_logits and llm_req.is_last_context_chunk:
                llm_req.py_result.append_generation_logits(logits_view)

            # TODO: Implement this once we have beam width support
            # Scatter the output logits to the decoderLogits
            # req_beam_width = llm_req.get_beam_width_by_iter()
            # if req_beam_width > 1:
            #     # Tile logits of context requests
            #     logits_shape = logits_view.shape
            #     logits_type = logits_view.dtype
            #     # decoder_logits = buffer_manager.gpu((req_beam_width, logits_shape[1]), logits_type)
            #     # tensorrt_llm.runtime.kernels.tile_tensor(decoder_logits, logits_view, req_beam_width, stream)
            #     decoder_logits = decoder_logits.unsqueeze(0)
            # else:
            #     decoder_buffer_logits[seq_slot] = logits_view[:logits_view.shape[0], :1, :logits_view.shape[1]]

        # Needs to be done in bulk for the copy to work
        decoder_buffers.logits = decoder_buffer_logits

        return logits_index
